scripts/metadata_comparison/metadata_comparison/extractor.py

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python3
+#
+# extractor.py
+#
+# Purpose: Read workflow metadata from Cromwell, and all metadata for its jobs,
+# and upload it to a GCS bucket
+#
+# Usage: python3 extractor.py <GCS path> <workflowId> [<workflowId2> [...]]
+#
+# Python Prereqs (at least, the ones which I needed to manually install... YMMV):
+#
+#   * pip3 install --upgrade requests
+#   * pip3 install --upgrade google-api-python-client
+#   * pip3 install --upgrade google-cloud
+#   * pip3 install --upgrade google-cloud-storage
+#
+# Remember to login to create application default credentials before use:
+#   % gcloud auth application-default login
+
+import argparse
+import re
+import json
+import requests
+from google.cloud import storage
+import google.auth
+from googleapiclient.discovery import build as google_client_build
+
+def workflow_regex_validator(value):
+    """Makes sure that a value is a valid Cromwell workflow ID then returns the workflow ID"""
+    workflow_regex=re.compile('^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$')
+    if not workflow_regex.match(value):
+        msg = f'Invalid workflow ID {value}. Expected {workflow_regex.pattern}'
+        raise argparse.ArgumentTypeError(msg)
+    else:
+        return value
+
+
+def url_regex_validator(value):
+    """
+    Validates then extract the root of the Cromwell URL from the various URL strings which might be provided.
+    Deliberately flexible because it's tedious to remember which script requires which type of format.
+    eg:
+        'http://localhost' => 'http://localhost'
+        'http://localhost:8000' => 'http://localhost:8000'
+        'http://localhost:8000/' => 'http://localhost:8000'
+        'http://localhost:8000/api/workflows/' => 'http://localhost:8000'
+        'http://localhost:8000/custom/prefix/api/workflows/' => 'http://localhost:8000/custom/prefix'
+    """
+    url_regex = re.compile('(http(s?)://((?!/api).)*[^/])(/(api.*)?)?')
+    m = url_regex.match(value)
+    if m:
+        return m.group(1)
+    else:
+        msg = f'Invalid Cromwell URL {value}. Expected {url_regex.pattern}'
+        raise argparse.ArgumentTypeError(msg)
+
+
+def gcs_path_regex_validator(value):
+    """
+    Validates then extracts the bucket and object-path from a GS string. Returned as a pair.
+    eg:
+        'gs://bucket/path/to/directory/' -> ('bucket', 'path/to/directory')
+    """
+    gcs_regex = re.compile('^gs://([a-zA-Z0-9-]+)/(([a-zA-Z0-9-]+/)*[a-zA-Z0-9-]+)/?$')
+    m = gcs_regex.match(value)
+    if m:
+        return m.group(1), m.group(2)
+    else:
+        msg = f'Invalid GCS path {value}. Expected {gcs_regex.pattern}'
+        raise argparse.ArgumentTypeError(msg)
+
+
+def get_operation_id_number(value):
+    """
+    Validates then extracts from PAPI operation IDs just the final number.
+    eg:
+        'projects/project_name/operations/01234567891011121314' -> '01234567891011121314'
```

Comment:
TOL: Would it also work for the v2beta operation id format? Or we only need support for the old format?

Reply:
IMHO we should support both. I think this code would work for both but a test to confirm that would be nice. 🙂

Reply:
Ah, you're right. I know the v1 IDs are different too, so maybe this is the way to distinguish (rather than guessing based on backend names).

I'll make the changes in an upcoming part 3 PR (supporting all the different API versions) since there are also other reasons why v1 and v2beta wouldn't yet work